File: hashing/find_longest_substring.java
public int findLongestSubstring(String s, int k)
{
    Map<Character, Integer> counts = new HashMap<>();

    int left {};
    int ans {};

    for (int right = 0; right < s.length(); right++)
    {
        char c = s.charAt(right);
        counts.put(c, counts.getOrDefault(c, 0) + 1);

        while (counts.size() > k)
        {
            char remove = s.charAt(left);
            counts.put(remove, counts.getOrDefault(remove, 0) - 1);

            if (counts.get(remove) == 0)
            {
                counts.remove(remove);
            }
            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
}

File: hashing/find_longest_substring.js
let findLongestSubstring = (s, k) => {
    let counts = new Map();
    let left = 0, ans = 0;

    for (let right = 0; right < s.length; right++)
    {
        counts.set(s[right], (counts.get(s[right]) || 0) + 1);
        while (counts.size > k)
        {
            counts.set(s[left], counts.get(s[left]) - 1);
            if (counts.get(s[left]) == 0)
            {
                counts.delete(s[left]);
            }
            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
}